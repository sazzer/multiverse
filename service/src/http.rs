pub(crate) mod link;
pub(crate) mod pagination;
pub(crate) mod patch;
pub(crate) mod problem;
pub(crate) mod sorts;
