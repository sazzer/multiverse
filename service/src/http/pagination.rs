use crate::model::Pagination;
use rocket::request::{FromQuery, Query};
use std::{ops::Deref, str::FromStr};

/// Request details to describe the pagination of a request
#[derive(Debug)]
pub struct PaginationRequest(Pagination);

impl Deref for PaginationRequest {
    type Target = Pagination;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'q> FromQuery<'q> for PaginationRequest {
    /// No error is ever returned
    type Error = ();

    /// Load the pagination details from the request and return an appropriate struct.
    ///
    /// This always uses standard querystring parameters for the values, namely:
    /// - `offset`
    /// - `count`
    ///
    /// If these are not present then defaults will be used instead, as defined in the implementation of
    /// the `Default` trait for the `PaginationRequest` trait
    ///
    /// # Parameters
    /// - `query` - The query parameters to get the parameters from
    ///
    /// # Returns
    /// The pagination details
    fn from_query(query: Query<'q>) -> Result<Self, Self::Error> {
        let mut result = Pagination::default();

        for item in query {
            if item.key == "offset" {
                if let Ok(value) = u64::from_str(item.value.as_str()) {
                    result.offset = value;
                }
            } else if item.key == "count" {
                if let Ok(value) = u64::from_str(item.value.as_str()) {
                    result.count = value;
                }
            }
        }

        Ok(PaginationRequest(result))
    }
}
