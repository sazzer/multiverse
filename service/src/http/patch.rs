use serde::{Deserialize, Deserializer};

/// Enumeration of possible states for each field in an HTTP Patch
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Patch<T> {
    /// The field was missing - i.e. it was undefined
    Missing,
    // The field was present but didn't have a value
    Null,
    /// The field was present and had a value
    Value(T),
}

impl<T> Patch<T> {
    /// Map the value inside the Patch to a different value, possibly of a different type.
    ///
    /// A `Missing` or `Null` value is left as-is.
    /// A `Value(v)` will get transformed into the new value
    ///
    /// # Parameters
    /// - `f` - The function to transform the value
    ///
    /// # Returns
    /// The transformed value
    pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Patch<U> {
        match self {
            Patch::Missing => Patch::Missing,
            Patch::Null => Patch::Null,
            Patch::Value(v) => Patch::Value(f(v)),
        }
    }

    /// Filter the value inside the Patch, returning `Null` if the value fails the predicate
    ///
    /// A `Missing` or `Null` value is left as-is.
    /// A `Value(v)` will get transformed into the new value
    ///
    /// # Parameters
    /// - `predicate` - The function to evaluate the value
    ///
    /// # Returns
    /// The filtered value
    pub fn filter_null<P: FnOnce(&T) -> bool>(self, predicate: P) -> Self {
        match self {
            Patch::Missing => Patch::Missing,
            Patch::Null => Patch::Null,
            Patch::Value(v) => {
                if predicate(&v) {
                    Patch::Value(v)
                } else {
                    Patch::Null
                }
            }
        }
    }

    /// Convert the value to an `Option<T>` but wrapped in a `Result`.
    /// The `Result` returns `Err(err)` if the `Patch` is `Null`, otherwise returns an `Ok(Option<T>)`
    /// containing the value.
    ///
    /// # Parameters
    /// - `err` - The error value to use
    ///
    /// # Returns
    /// - `Patch::Value(v)` -> `Ok(Some(v))`
    /// - `Patch::Missing` -> `Ok(None`
    /// - `Patch::Null` -> `Err(err)`
    pub fn ok_not_null<E>(self, err: E) -> Result<Option<T>, E> {
        match self {
            Patch::Null => Err(err),
            Patch::Missing => Ok(None),
            Patch::Value(v) => Ok(Some(v)),
        }
    }
}

impl<T, E> Patch<Result<T, E>> {
    /// Transpose a patch containing a result into a result containing the patch
    ///
    /// # Returns
    /// - `Patch::Missing` -> Ok(Missing)
    /// - `Patch::Null` -> Ok(Null)
    /// - `Patch::Value(Ok(v)) -> Ok(Value(v))
    /// - `Patch::Value(Err(e)) -> Err(e)
    pub fn transpose(self) -> Result<Patch<T>, E> {
        match self {
            Patch::Missing => Ok(Patch::Missing),
            Patch::Null => Ok(Patch::Null),
            Patch::Value(Ok(v)) => Ok(Patch::Value(v)),
            Patch::Value(Err(e)) => Err(e),
        }
    }
}

impl<T> Default for Patch<T> {
    fn default() -> Self {
        Patch::Missing
    }
}

impl<T> From<Option<T>> for Patch<T> {
    fn from(opt: Option<T>) -> Patch<T> {
        match opt {
            Some(v) => Patch::Value(v),
            None => Patch::Null,
        }
    }
}

impl<'de, T> Deserialize<'de> for Patch<T>
where
    T: Deserialize<'de>,
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Option::deserialize(deserializer).map(Into::into)
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    use galvanic_assert::{assert_that, matchers::*};

    #[test]
    fn test_map() {
        assert_that!(&Patch::Missing.map(|v: u32| v), eq(Patch::Missing));
        assert_that!(&Patch::Null.map(|v: u32| v), eq(Patch::Null));
        assert_that!(&Patch::Value(5).map(|v| v * 2), eq(Patch::Value(10)));
        assert_that!(
            &Patch::Value(5).map(|v| format!("{}", v)),
            eq(Patch::Value("5".to_owned()))
        );
    }

    #[test]
    fn test_filter_null() {
        assert_that!(
            &Patch::Missing.filter_null(|_: &u32| true),
            eq(Patch::Missing)
        );
        assert_that!(
            &Patch::Missing.filter_null(|_: &u32| false),
            eq(Patch::Missing)
        );
        assert_that!(&Patch::Null.filter_null(|_: &u32| true), eq(Patch::Null));
        assert_that!(&Patch::Null.filter_null(|_: &u32| false), eq(Patch::Null));
        assert_that!(&Patch::Value(1).filter_null(|_| true), eq(Patch::Value(1)));
        assert_that!(&Patch::Value(1).filter_null(|_| false), eq(Patch::Null));
    }

    #[test]
    fn test_ok_not_null() {
        assert_that!(&Patch::<u32>::Missing.ok_not_null(5), eq(Ok(None)));
        assert_that!(&Patch::<u32>::Null.ok_not_null(5), eq(Err(5)));
        assert_that!(&Patch::<u32>::Value(1).ok_not_null(5), eq(Ok(Some(1))));
    }

    #[test]
    fn test_transpose() {
        assert_that!(
            &Patch::<Result<u32, u32>>::Missing.transpose(),
            eq(Ok(Patch::Missing))
        );
        assert_that!(
            &Patch::<Result<u32, u32>>::Null.transpose(),
            eq(Ok(Patch::Null))
        );
        assert_that!(
            &Patch::<Result<u32, u32>>::Value(Ok(1)).transpose(),
            eq(Ok(Patch::Value(1)))
        );
        assert_that!(
            &Patch::<Result<u32, u32>>::Value(Err(5)).transpose(),
            eq(Err(5))
        );
    }
}
