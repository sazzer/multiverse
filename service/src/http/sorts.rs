use crate::model::{ParseSortFieldError, SortFields};
use rocket::{http::RawStr, request::FromFormValue};
use std::{ops::Deref, str::FromStr};

/// Request details to describe the sort fields of a request
#[derive(Debug)]
pub struct SortFieldsRequest<T>(SortFields<T>);

impl<T> Deref for SortFieldsRequest<T> {
    type Target = SortFields<T>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'v, T> FromFormValue<'v> for SortFieldsRequest<T>
where
    T: FromStr,
    T: strum::VariantNames,
    T: std::fmt::Debug,
{
    type Error = ParseSortFieldError;

    fn from_form_value(form_value: &'v RawStr) -> Result<Self, Self::Error> {
        let sort_fields: SortFields<T> = form_value.as_str().parse()?;

        Ok(SortFieldsRequest(sort_fields))
    }

    fn default() -> Option<Self> {
        Some(SortFieldsRequest(SortFields(vec![])))
    }
}
