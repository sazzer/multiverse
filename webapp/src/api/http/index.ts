export { Problem } from "./problem";
export { request } from "./request";
