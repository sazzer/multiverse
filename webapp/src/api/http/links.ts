import debug from "debug";

/** The logger to use */
const LOGGER = debug("multiverse:api:http:links");

/** Regex to extract the link target */
const LINK_TARGET_REGEX = /<(.+)>(.*)/;

/** The shape of a link parameter */
export interface LinkParameter {
  /** The parameter key */
  key: string;

  /** The parameter value */
  value: string;
}

/** The shape of a link */
export interface Link {
  /** The target of the link */
  target: string;

  /** The */
  parameters: LinkParameter[];
}

/** Representation of a set of links */
export class Links {
  /** The actual links */
  readonly links: Link[];

  /**
   * Construct the links
   * @param links The links
   */
  constructor(links: Link[]) {
    this.links = links;
  }

  /**
   * Get the link that has the given link relation
   * @param rel The link relation to look for
   */
  getLinkByRel(rel: string): Link | undefined {
    return this.links.find((link) =>
      link.parameters.find(
        (parameter) => parameter.key === "rel" && parameter.value === rel
      )
    );
  }
}

/**
 * Parse a single link value from the link header
 * @param value The string to parse
 */
export function parseLink(value: string): Link {
  const match = value.match(LINK_TARGET_REGEX);
  if (match) {
    const target = match[1];
    const parameters = match[2]
      .split(";")
      .map((param) => param.trim())
      .filter((param) => param.length > 0)
      .map((param) => param.split("=", 2))
      .filter((param) => param.length === 2)
      .map(([key, value]) => {
        let trimmedValue = value.trim();
        if (trimmedValue.startsWith('"') && trimmedValue.endsWith('"')) {
          trimmedValue = trimmedValue.substr(1, trimmedValue.length - 2);
        }
        return {
          key: key.trim(),
          value: trimmedValue,
        };
      });
    return { target, parameters };
  } else {
    throw new Error("Malformed link");
  }
}

/**
 * Parse an entire set of linsk from the link header
 * @param value The value to parse
 */
export function parseLinks(value: string): Links {
  const links = value
    .split(",")
    .map((link) => link.trim())
    .filter((link) => link.length > 0)
    .map(parseLink);
  LOGGER("Parsed links: %o", links);
  return new Links(links);
}
