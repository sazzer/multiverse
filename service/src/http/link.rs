use rocket::http::Header;
use std::fmt::{Display, Formatter};

/// Possible Link relations
#[derive(Debug, PartialEq)]
pub enum LinkRel {
    /// A link back to the resource
    SelfLink,
    /// A link to a related resource
    Related,
    /// Any custom link relation
    Custom(String),
}

/// Representation of a Link header
#[derive(Debug)]
pub struct Link {
    /// The target of the link
    target: String,
    /// The link relation
    rel: LinkRel,
}

/// Representation of a set of link headers
pub struct Links(pub Vec<Link>);

impl Link {
    /// Create a new Link header
    ///
    /// # Parameters
    /// - `target` - The target of the link
    /// - `rel` - The link relation
    ///
    /// # Returns
    /// The link header
    pub fn new<T>(target: T, rel: LinkRel) -> Self
    where
        T: Into<String>,
    {
        Self {
            target: target.into(),
            rel,
        }
    }
}

impl Display for LinkRel {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let relation = match self {
            LinkRel::SelfLink => "self",
            LinkRel::Related => "related",
            LinkRel::Custom(rel) => rel.as_ref(),
        };
        write!(f, "rel=\"{}\"", relation)
    }
}

impl Display for Link {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "<{}>; {}", self.target, self.rel)
    }
}

impl<'h> From<Link> for Header<'h> {
    fn from(link: Link) -> Self {
        Header::new("Link", format!("{}", link))
    }
}

impl<'h> From<Links> for Header<'h> {
    fn from(links: Links) -> Self {
        let output: Vec<String> = links.0.iter().map(|link| format!("{}", link)).collect();
        Header::new("Link", output.join(","))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use galvanic_assert::{assert_that, matchers::*};

    #[test]
    pub fn test_format_self_link() {
        let link = Link::new("/example", LinkRel::SelfLink);
        let formatted = format!("{}", link);

        assert_that!(&formatted, eq("</example>; rel=\"self\"".to_owned()));
    }

    #[test]
    pub fn test_format_custom_link() {
        let link = Link::new("/example", LinkRel::Custom("somethingElse".to_owned()));
        let formatted = format!("{}", link);

        assert_that!(
            &formatted,
            eq("</example>; rel=\"somethingElse\"".to_owned())
        );
    }

    #[test]
    pub fn test_build_single_link_header() {
        let link = Link::new("/example", LinkRel::SelfLink);
        let header: Header<'_> = link.into();
        let formatted = format!("{}", header);

        assert_that!(&formatted, eq("Link: </example>; rel=\"self\"".to_owned()));
    }

    #[test]
    pub fn test_build_multiple_link_header() {
        let links = Links(vec![
            Link::new("/example1", LinkRel::SelfLink),
            Link::new("/example2", LinkRel::SelfLink),
        ]);

        let header: Header<'_> = links.into();
        let formatted = format!("{}", header);

        assert_that!(
            &formatted,
            eq("Link: </example1>; rel=\"self\",</example2>; rel=\"self\"".to_owned())
        );
    }
}
