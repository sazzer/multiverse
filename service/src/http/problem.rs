use rocket::http::Status;
use serde::Serialize;
use serde_json::Value;
use std::{
    collections::HashMap,
    fmt::{Debug, Display, Formatter},
};
pub use validation::*;

mod response;
mod validation;

/// Trait to represent the type of a problem
pub trait ProblemType: Display + Debug {
    /// Generate a Type value for the `ProblemType` values.
    ///
    /// These are used in the `type` field in the RFC-7807 Problem Response
    fn error_code(&self) -> &'static str;
}

/// Representation of a Problem expressed in terms of RFC-7807
#[derive(Debug)]
pub struct Problem {
    /// The actual error that occurred
    pub error: Box<dyn ProblemType>,
    /// The HTTP Status code to use
    pub status: Status,
    /// An additional detail message
    pub detail: Option<String>,
    /// An additional instance subtype
    pub instance: Option<String>,
    /// Any extra details
    pub extra: HashMap<String, Value>,
}

impl Problem {
    /// Create a new Problem instance
    ///
    /// # Parameters
    /// - `error` - The error code
    /// - `status` - The HTTP Status code
    ///
    /// # Returns
    /// The problem
    pub fn new<T>(error: T, status: Status) -> Self
    where
        T: ProblemType + 'static,
    {
        Self {
            error: Box::new(error),
            status,
            detail: None,
            instance: None,
            extra: HashMap::new(),
        }
    }

    /// Set the Detail of the Problem instance
    #[allow(dead_code)]
    pub fn with_detail<S>(self, detail: S) -> Self
    where
        S: Into<String>,
    {
        Self {
            detail: Some(detail.into()),
            ..self
        }
    }

    /// Set the Instance of the Problem instance
    #[allow(dead_code)]
    pub fn with_instance<S>(self, instance: S) -> Self
    where
        S: Into<String>,
    {
        Self {
            instance: Some(instance.into()),
            ..self
        }
    }

    /// Set some extra data on the Problem instance
    pub fn with_extra<K, V>(self, key: K, value: V) -> Self
    where
        K: Into<String>,
        V: Serialize,
    {
        let mut extra = self.extra;
        extra.insert(
            key.into(),
            serde_json::to_value(value).expect("Failed to serialize extra detail"),
        );

        Self { extra, ..self }
    }
}

impl Display for Problem {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.error)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(thiserror::Error, Debug, PartialEq)]
    pub enum ProblemDetails {
        #[error("Something Happened")]
        SomeProblem,
    }

    impl ProblemType for ProblemDetails {
        fn error_code(&self) -> &'static str {
            "tag:spacegame,2020:some/problem"
        }
    }

    #[test]
    fn test_basic_problem() {
        let problem = Problem::new(ProblemDetails::SomeProblem, Status::BadRequest);

        assert_eq!(Status::BadRequest, problem.status);
        // TODO: assert_matches!(*problem.error, ProblemDetails::SomeProblem);
        assert_eq!(None, problem.detail);
        assert_eq!(None, problem.instance);
        assert_eq!(0, problem.extra.len());
    }

    #[test]
    fn test_full_problem() {
        let problem = Problem::new(ProblemDetails::SomeProblem, Status::BadRequest)
            .with_detail("Some Detail")
            .with_instance("Some Instance")
            .with_extra("some_key", "Some Value")
            .with_extra("other_key", 42);

        assert_eq!(Status::BadRequest, problem.status);
        // TODO: assert_matches!(*problem.error, ProblemDetails::SomeProblem);
        assert_eq!(Some("Some Detail".to_owned()), problem.detail);
        assert_eq!(Some("Some Instance".to_owned()), problem.instance);
        assert_eq!(2, problem.extra.len());
        assert_eq!(
            Some(&serde_json::to_value("Some Value").unwrap()),
            problem.extra.get(&"some_key".to_owned())
        );
        assert_eq!(
            Some(&serde_json::to_value(42).unwrap()),
            problem.extra.get(&"other_key".to_owned())
        );
    }

    #[test]
    fn test_problem_display() {
        let problem = Problem::new(ProblemDetails::SomeProblem, Status::BadRequest);

        assert_eq!("Something Happened".to_owned(), format!("{}", problem));
    }
}
