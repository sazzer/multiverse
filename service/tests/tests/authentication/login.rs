use crate::{
    data::{hash_password, SeedUser},
    tests::run_test,
};
use insta::assert_json_snapshot;
use rocket::http::Status;
use serde_json::json;

#[test]
fn test_login_unknown_user() {
    run_test()
        .post(
            "/login",
            json!({
                "username": "username",
                "password": "password"
            }),
        )
        .has_status(Status::Unauthorized)
        .has_header("Content-Type", "application/problem+json")
        .has_json_body(json!({
            "type": "tag:multiverse,2020:users/problems/authentication_error",
            "title": "Invalid Username or Password",
            "status": 401
        }));
}

#[test]
fn test_login_empty_body() {
    run_test()
        .post("/login", json!({}))
        .has_status(Status::Unauthorized)
        .has_header("Content-Type", "application/problem+json")
        .has_json_body(json!({
            "type": "tag:multiverse,2020:users/problems/authentication_error",
            "title": "Invalid Username or Password",
            "status": 401
        }));
}

#[test]
fn test_login_invalid_password() {
    run_test()
        .seed(SeedUser {
            username: "testuser".to_owned(),
            password: hash_password("password"),
            ..SeedUser::default()
        })
        .post(
            "/login",
            json!({
                "username": "testuser",
                "password": "incorrect"
            }),
        )
        .has_status(Status::Unauthorized)
        .has_header("Content-Type", "application/problem+json")
        .has_json_body(json!({
            "type": "tag:multiverse,2020:users/problems/authentication_error",
            "title": "Invalid Username or Password",
            "status": 401
        }));
}

#[test]
fn test_login_missing_password() {
    run_test()
        .seed(SeedUser {
            username: "testuser".to_owned(),
            password: hash_password("password"),
            ..SeedUser::default()
        })
        .post(
            "/login",
            json!({
                "username": "testuser"
            }),
        )
        .has_status(Status::Unauthorized)
        .has_header("Content-Type", "application/problem+json")
        .has_json_body(json!({
            "type": "tag:multiverse,2020:users/problems/authentication_error",
            "title": "Invalid Username or Password",
            "status": 401
        }));
}

#[test]
fn test_login_success() {
    run_test()
        .seed(SeedUser {
            username: "testuser".to_owned(),
            password: hash_password("password"),
            display_name: "Test User".to_owned(),
            email_address: "testuser@example.com".to_owned(),
            ..SeedUser::default()
        })
        .post(
            "/login",
            json!({
                "username": "testuser",
                "password": "password"
            }),
        )
        .has_status(Status::Ok)
        .has_header("Content-Type", "application/json")
        .assert_json_body(|body| {
            assert_json_snapshot!(body, {
                ".token.token" => "[access_token]",
                ".token.valid_until" => "[access_token_expiry]"
            }, @r###"
            {
              "token": {
                "token": "[access_token]",
                "valid_until": "[access_token_expiry]"
              },
              "user": {
                "display_name": "Test User",
                "email_address": "testuser@example.com",
                "username": "testuser"
              }
            }
            "###);
        });
}
