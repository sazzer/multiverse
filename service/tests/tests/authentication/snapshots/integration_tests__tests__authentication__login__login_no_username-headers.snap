---
source: tests/tests/authentication/login.rs
expression: response.headers()
---
"HTTP/1.1 401 Unauthorized\nContent-Type: application/problem+json\nServer: Rocket"
